scheduler/server/task_scheduler.go

```diff
@@ -81,48 +82,54 @@ func (s *statefulScheduler) getTaskAssignments() []taskAssignment {
 	return assignments
 }
 
+type nodeStatesByNodeID map[cluster.NodeId]*nodeState
+
 // Helper fn, appends to 'assignments' and updates nodeGroups.
 // Should successfully assign all given tasks if caller invokes this with self-consistent params.
-func assign(
-	cs *clusterState,
-	tasks []*taskState,
-	stat stats.StatsReceiver,
-) (assignments []taskAssignment) {
-	idleNodeGroupIDs := map[string]string{}
-	for groupID, group := range cs.nodeGroups {
+// Note: there may be a race condition between recognizing idle nodes as available for assignment and
+// nodes becoming offlined or suspended. The code does the best it can, but it may assign a task to
+// a node that clusterState considers offlined/suspended before or as the task is actually being started
+func (s *statefulScheduler) assign(tasks []*taskState) (assignments []taskAssignment) {
+	idleNodesByGroupIDs := map[string]nodeStatesByNodeID{}
+
+	// make a local copy of (non-suspended/offlined) idle nodeIDs by groupID, these will be the nodes assigned to the tasks
```

Comment:
why do we need a local copy? can we include that reasoning in this comment?

Reply:
good idea - this tripped me up for a while....